### DISCUSSION TITLE
Distribute type definitions in the `react-native` package and phase out `@types/react-native`

### DISCUSSION THREAD

**Discussion by tom-sherman** [react-native-community/discussions-and-proposals#490] [Ideas]
I'm raising this after we forgot to bump `@types/react-native` in our app and it caused a bug. The bug was autocomplete broke after we upgraded to 0.66, because that release removed `autoCompleteType` on Android in favour of `autoComplete`.

The root cause here IMO is that developers need to be careful when upgrading React Native that they also upgrade the types package.

The preferred solution here, which eliminates the problem entirely, is to distribute the types in official React Native releases.

I understand why this wasn't done historically, TypeScript wasn't as widespread as it is now. However I would guess (and if anyone has data please share!) a vast majority of apps are written in TypeScript and could benefit from this change.

**Comment by kelset** [comment]
Problem is, react-native is typed with Flow because that's what is used within the Meta monorepo.

So TS types can't be added to the react-native node_module and the best thing you can do is instead submit PRs to the DefinitelyTyped repo to address the issues you face.

**Reply by kelset** [reply]
For context, I've started submitting PRs myself to create the new react-native minor releases: https://github.com/DefinitelyTyped/DefinitelyTyped/pulls?q=is%3Apr+author%3A%40me+sort%3Aupdated-desc+is%3Aclosed

**Reply by tom-sherman** [reply]
As far as I'm aware there's nothing stopping `react-native.d.ts` to be developed and released inside of the official package. You can have both flow and typescript definitions side by side.

I'm not saying convert React Native internals to typescript, just develop and distribute the declarations in the official package instead of via DefinitelyTyped

**Reply by kelset** [reply]
> nothing stopping react-native.d.ts to be developed and released inside of the official package

Maybe I should have been clearer: because react-native is a mirror of the code that lives within the Meta monorepo, there can't be .ts files in there.

**Reply by kelset** [reply]
Also, having types in the official package implies that someone will have to maintain them - and who would that be? The Meta engineers won't because again, Meta monorepo uses Flow; so it must be someone else, but at that point the DT repo is a much smoother experience for doing so.

Again, because of how things are setup right now, the best thing you can do is submit PRs to the DefinitelyTyped repo to address the issues you face. There's really good automation going on there and usually PRs get merged fairly quickly.

**Reply by kelset** [reply]
(all of that said: we at MSFT know that having good TS types is important for RN and we care a lot about this, so we'll be trying to help out more ensuring that the types are in good shape - we have in particular one engineer that will work on this more for the next few months)

**Comment by lunaleaps** [comment]
We've done just this! We will be deprecating @types/react-native from 0.72 onward. See changes here: https://github.com/facebook/react-native/commit/6b2a511cbb1f244133624e37e7b3215d8fed33af

These changes will be in the 0.71 release so you can remove the @types/react-native requirement for anyone using 0.71
